model/execution/DeliveryExecutionManagerService.php

```diff
@@ -106,12 +109,44 @@ public function setExtraTime($deliveryExecutions, $extraTime = null, $extendedTi
                 $testDefinition = \taoQtiTest_helpers_Utils::getTestDefinition($inputParameters['QtiTestCompilation']);
                 $deliveryExecutionArray[] = $deliveryExecution;
                 $extraTime = null;
+                $seconds = null;
+
                 if ($testDefinition->getTimeLimits() && $testDefinition->getTimeLimits()->hasMaxTime()) {
                     $maxTime = $testDefinition->getTimeLimits()->getMaxTime();
                     $seconds = $maxTime->getSeconds(true);
+                }
+
+                /** @var TestSession $testSession */
+                if ($testSession = $testSessionService->getTestSession($deliveryExecution)) {
+                    if ($testPart = $testSession->getCurrentTestPart()) {
+                        if ($testSessionLimits = $testPart->getTimeLimits()) {
+                            $seconds = $testSessionLimits->hasMaxTime()
+                                ? $testSessionLimits->getMaxTime()->getSeconds(true)
+                                : $seconds;
+                        }
+                    }
+
+                    if ($section = $testSession->getCurrentAssessmentSection()) {
+                        if ($testSessionLimits = $section->getTimeLimits()) {
+                            $seconds = $testSessionLimits->hasMaxTime()
+                                ? $testSessionLimits->getMaxTime()->getSeconds(true)
+                                : $seconds;
+                        }
+                    }
+
+                    if ($item = $testSession->getCurrentAssessmentItemRef()) {
```

Comment:
It's better to check items timers first and then if `$seconds` is `null` search it in upper elements